### ISSUE TITLE
Advent of cats-parse

### ISSUE THREAD

**Issue by rossabaker** [http4s/http4s#3984]
These files have parboiled imports that need to go.  We are moving them to cats-parse.  Some of these files are related. Some are enormous and can be broken down.  They could be tasked better, but this is a start.

All work should target the series-0.22 branch, but all of this will get it closer.

Basic steps:

1. We are putting [parsers on the companion objects](https://github.com/http4s/http4s/blob/79ed398dfe6da5988e7752b0943c16b749736f38/core/src/main/scala/org/http4s/HttpDate.scala#L127-L130) of the type they parse, as `private[http4s]`.  By being on the companion, they can invoke private constructors with validated values.  Commenting on the RFC rules each parser implements is nice, as it makes it easier to trace back to the spec.

2. Replace the `.fromString` or `.parse` method on the companion with [code like this](https://github.com/http4s/http4s/blob/79ed398dfe6da5988e7752b0943c16b749736f38/core/src/main/scala/org/http4s/HttpDate.scala#L83-L86).

3. Delete any parboiled parsers left behind.

4. Run `core/mimaReportBinaryIssues` and [add MiMa exclusions](https://github.com/http4s/http4s/blob/dotty/build.sbt#L101).  They were all private, so it's safe.

5. If you find common rules, we are [putting some RFC objects](https://github.com/http4s/http4s/tree/dotty/core/src/main/scala/org/http4s/internal/parsing) in an internal package to share.

6. Some of these are trivial and some are very hard.  If you get in over your head, [we're on Gitter](https://gitter.im/http4s/http4s-dev) or we're here.

If you're interested in one, please comment:

* [x] ./core/src/main/scala/org/http4s/ContentCoding.scala (@novakov-alexey)
* [x] ./core/src/main/scala/org/http4s/MediaType.scala (@novakov-alexey)
* [x] ./core/src/main/scala/org/http4s/QValue.scala (@novakov-alexey)
* [x] ./core/src/main/scala/org/http4s/Query.scala (@rossabaker)
* [x] ./core/src/main/scala/org/http4s/TransferCoding.scala (#4023)
* [x] ./core/src/main/scala/org/http4s/Uri.scala (@rossabaker)
* [x] ./core/src/main/scala/org/http4s/parser/AcceptCharsetHeader.scala (@novakov-alexey)
* [x] ./core/src/main/scala/org/http4s/parser/AcceptEncodingHeader.scala (@novakov-alexey)
* [x] ./core/src/main/scala/org/http4s/parser/AcceptHeader.scala (@novakov-alexey)
* [x] ./core/src/main/scala/org/http4s/parser/AcceptLanguageHeader.scala (@novakov-alexey)
* [x] ./core/src/main/scala/org/http4s/parser/AdditionalRules.scala
* [x] ./core/src/main/scala/org/http4s/parser/AuthorizationHeader.scala (@hamnis)
* [x] ./core/src/main/scala/org/http4s/parser/CacheControlHeader.scala (@deniszo)
* [x] ./core/src/main/scala/org/http4s/parser/ChallengeParser.scala (@hamnis)
* [x] ./core/src/main/scala/org/http4s/parser/ContentTypeHeader.scala (@novakov-alexey)
* [x] ./core/src/main/scala/org/http4s/parser/CookieHeader.scala (@rossabaker)
* [x] ./core/src/main/scala/org/http4s/parser/ForwardedHeader.scala (@fredshonorio)
* [x] ./core/src/main/scala/org/http4s/parser/ForwardedModelParsing.scala (@fredshonorio)
* [x] ./core/src/main/scala/org/http4s/parser/Http4sHeaderParser.scala
* [x] ./core/src/main/scala/org/http4s/parser/Http4sParser.scala
* [x] ./core/src/main/scala/org/http4s/parser/IpParser.scala (#4004)
* [x] ./core/src/main/scala/org/http4s/parser/LinkHeader.scala (@novakov-alexey)
* [x] ./core/src/main/scala/org/http4s/parser/OriginHeader.scala (@lewisjkl)
* [x] ./core/src/main/scala/org/http4s/parser/ProxyAuthenticateHeader.scala (@hamnis)
* [x] ./core/src/main/scala/org/http4s/parser/RangeParser.scala (@fredshonorio)
* [x] ./core/src/main/scala/org/http4s/parser/RequestUriParser.scala (@rossabaker)
* [x] ./core/src/main/scala/org/http4s/parser/Rfc2616BasicRules.scala
* [x] ./core/src/main/scala/org/http4s/parser/Rfc3986Parser.scala (@rossabaker)
* [x] ./core/src/main/scala/org/http4s/parser/SimpleHeaders.scala
* [x] ./core/src/main/scala/org/http4s/parser/StrictTransportSecurityParser.scala
* [x] ./core/src/main/scala/org/http4s/parser/UriHeaderParser.scala
* [x] ./core/src/main/scala/org/http4s/parser/WwwAuthenticateHeader.scala (@hamnis)
* [x] ./core/src/main/scala/org/http4s/parser/ZipkinHeader.scala (@m-sp)
* [x] ./core/src/main/scala/org/http4s/util/UrlCoding.scala (@rossabaker)

Inspired by [this Twitter thread](https://twitter.com/bjoernmadsen/status/1335284254746550272).

**Comment by lewisjkl** [comment]
Hey I'd love to help out with this! I can start with OriginHeader and then potentially pick up some more after getting that PR submitted/merged.

**Comment by deniszo** [comment]
Going to a take a shot at `ContentTypeHeader` parser

**Comment by fredshonorio** [comment]
I'll go for RangeParser

**Comment by rossabaker** [comment]
Thanks for volunteering!  Hopefully the guide above is useful.  We'll refine it as we get more people involved.

**Comment by deniszo** [comment]
After giving up and deleting my previous comment, I am back at taking a shot at `CacheControlHeader` 😅

**Comment by rossabaker** [comment]
:tada: Please don't be afraid to ask for advice if they're hard.  I haven't been very good myself at predicting which ones would be easy and which ones wake dragons.

**Comment by rossabaker** [comment]
(Administrative note: putting names next to ones in progress, but now using the check to reflect they're merged.)

**Comment by lewisjkl** [comment]
I started going through the OriginHeader last night and for the most part it is straight forward. One thing I realized though is that I am going to need a parser for ipv4 and ipv6 addresses in order to parse the host names out. I am happy to go ahead and write those, but I wanted to ask if those maybe belong in another location? Or maybe they're already done somewhere else?

**Comment by rossabaker** [comment]
@lewisjkl I've been putting common rules in the RFCs they come from.  Then, when an RFC references another, we can mimic that buy importing another RFC object.  So I'd put the IP address rules an `Rfc3986` object in `org.http4s.internals.parsing`.  URI will also benefit from those.

There are parboiled rules for these already, but I recall the IPv6 one being particularly ugly.

**Comment by fredshonorio** [comment]
I've added https://github.com/http4s/http4s/pull/3995, should I ping anyone?

**Comment by rossabaker** [comment]
Nope, we'll take a look. Thanks!

**Comment by hamnis** [comment]
I'm having a look at `Authorization`

**Comment by rossabaker** [comment]
@fredshonorio Thanks for the `Accept-Ranges` parser.  Do you want to keep going and finish off all the range parsers?

**Comment by fredshonorio** [comment]
> 
> 
> @fredshonorio Thanks for the `Accept-Ranges` parser. Do you want to keep going and finish off all the range parsers?

For sure!

**Comment by lewisjkl** [comment]
Hey all, just submitted a WIP PR for OriginHeader. I am a bit stuck with the ipv6 parser and would love any thoughts/ideas you all have. See #4004.

**Comment by hamnis** [comment]
I`ll pick up the rest of RFC7235, that is `WWW-Authenticate`, `Proxy-Authenticate` and `Proxy-Authorization`
Most of them are based on, or related to Authorization.

@rossabaker should I do all of them in the same PR? I currently have a branch based on the [Authorization PR](#4001)

**Comment by novakov-alexey** [comment]
Hi there. I will take:
./core/src/main/scala/org/http4s/QValue.scala
./core/src/main/scala/org/http4s/ContentCoding.scala

**Comment by novakov-alexey** [comment]
I will also cover:
./core/src/main/scala/org/http4s/parser/AcceptCharsetHeader.scala
./core/src/main/scala/org/http4s/parser/AcceptEncodingHeader.scala
./core/src/main/scala/org/http4s/parser/AcceptLanguageHeader.scala

**Comment by zmccoy** [comment]
@rossabaker It'd be good to link the https://github.com/http4s/http4s/pull/4004 PR to the `./core/src/main/scala/org/http4s/parser/IpParser.scala` line being that it requires the IPV6 parser that is done in there.  Happy I checked how people were doing `(1 to X).times` to find that before I went further 😅

**Comment by lewisjkl** [comment]
Re: #4004, I got busy the last few days so had to put it on hold. Should be able to hopefully get it completed tomorrow though since I finally got ipv6 parsing 🎉

**Comment by rossabaker** [comment]
@zmccoy Do you want to grab one? I'm not sure what's left of `IpParser` after #4004.

**Comment by fredshonorio** [comment]
Here's https://github.com/http4s/http4s/pull/4017

**Comment by novakov-alexey** [comment]
I am gonna take another one:
./core/src/main/scala/org/http4s/MediaType.scala

**Comment by novakov-alexey** [comment]
MediaType is used by these 3 headers, so I take them as well :-)
 ./core/src/main/scala/org/http4s/parser/AcceptHeader.scala
./core/src/main/scala/org/http4s/parser/ContentTypeHeader.scala
./core/src/main/scala/org/http4s/parser/LinkHeader.scala

**Comment by zmccoy** [comment]
@rossabaker Sorry I'm not able to commit right now, but if that changes I'll ping in here on what I'm taking.  :)

**Comment by fredshonorio** [comment]
I'll also take
./core/src/main/scala/org/http4s/parser/ForwardedHeader.scala
./core/src/main/scala/org/http4s/parser/ForwardedModelParsing.scala

**Comment by m-sp** [comment]
I'll take a look at the `ZipkinHeader` parsers

**Comment by rossabaker** [comment]
:tada: This was a massive effort. Thanks to all who participated!
